### DISCUSSION TITLE
PROTECTOR 48KW Controller and Pi not communicating

### DISCUSSION THREAD

**Discussion by EvanET13** [jgyates/genmon#1031] [Questions] [answered]
I am trying to connect my generator via a Pi 4B and Gen HAT from pintsize.me; when connecting the pi to the controller (via the accessory port) I am unable to see any stats on the web gui. It was consistently saying "initializing connection" or something along the lines. I believe it is a [PROTECTOR 48KW](https://www.generac.com/all-products/generators/commercial-generators/protector-series/48kw-rd04833); wasn't given any of the paperwork for it. But it does have the evolution controller.

When I run the test on the pi the gui load properly. But when connecting the controller to the pi, the only stats that are shown are those of the Pi (wifi signal and CPU temp).

What course of action can i take from here? I did try the troubleshoot page in the wiki but it did not help.

**Comment by skipfire** [comment]
Have you done a serial loopback test to confirm that serial is correctly enabled?  This needs to be done with the data lines disconnected from the generator and the GenMon application shut down.

**Comment by jgyates** [comment]
Did you order the HAT and setup the pi yourself? if you did setup the pi your self, did you run the installation script described here: https://github.com/jgyates/genmon/wiki/3.2--Setup-Your-Pi-and-Serial-Port

**Reply by EvanET13** [reply]
I've removed the HAT and just tried jumping the onboard serial (pins 8 and 10). I receive the following:
![image](https://github.com/jgyates/genmon/assets/60356533/67b5d5ff-8cbf-4912-8a7b-65c84b7eeac5)

I have also ran the serialconfig.py. In addition I did make sure to stop genmon. I also made sure to setup the pi to so onboard serial. Still having the same issues.

**Reply by EvanET13** [reply]
I did follow all the steps in https://github.com/jgyates/genmon/wiki/3.2--Setup-Your-Pi-and-Serial-Port making sure uart is enabled, etc.

**Comment by skipfire** [comment]
You may need to see what serial ports you have, `ls /dev/serial*`, you may need to update the config to use serial1 or a tty.

**Reply by EvanET13** [reply]
i thought the same but got these results:
![image](https://github.com/jgyates/genmon/assets/60356533/40ae51e1-5772-404a-8be0-b928a58e80ea)

**Comment by jgyates** [comment]
did you run this script?

https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software

The genmonmaint.sh script, with the -i option will install the needed libraries, copy the config files and setup the serial port and make the needed changes to the config file. If you have not done this then you are skipping about 3 steps that you will need to run genmon. Also, this page describes how to download genmon. Just making a copy manually from github will not allow updates to occur. Follow the directions on the link above. If that does not work then we can go from there.

**Reply by EvanET13** [reply]
I will try this again and follow up with my results.

**Reply by EvanET13** [reply]
Still having the issue. I removed the whole directory and started fresh. Still no luck. Attaching a code snippet of the entire process.

```
cvin-admin@genmon:~ $  git clone https://github.com/jgyates/genmon.git
Cloning into 'genmon'...
remote: Enumerating objects: 10470, done.
remote: Counting objects: 100% (3713/3713), done.
remote: Compressing objects: 100% (1070/1070), done.
remote: Total 10470 (delta 2754), reused 3458 (delta 2640), pack-reused 6757
Receiving objects: 100% (10470/10470), 48.29 MiB | 2.46 MiB/s, done.
Resolving deltas: 100% (7598/7598), done.
cvin-admin@genmon:~ $ cd genmon/
cvin-admin@genmon:~/genmon $ cd OtherApps/
cvin-admin@genmon:~/genmon/OtherApps $ python3 serialtest.py 

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5b0
Serial port opened
write data: sent test string
waiting to received data....
FAILED: Sent data does not match receive. Received 0 bytes
cvin-admin@genmon:~/genmon/OtherApps $ python3 serialconfig.py -c
You need to have root privileges to run this script.
Please try again, this time using 'sudo'. Exiting.
cvin-admin@genmon:~/genmon/OtherApps $ sudo python3 serialconfig.py -c

Model: Raspberry Pi 4 Model B Rev 1.5

Checking : Is enable UART in boot config: OK
Checking : Is BT overlay disabled: OK
Checking : Serial console command line removed: OK
Checking : Serial console service disabled: OK
Checking : BT service disabled: OK

Serial port settings are OK. A reboot is needed if changes were made.

cvin-admin@genmon:~/genmon/OtherApps $ cd ..
cvin-admin@genmon:~/genmon $ bash ./genmonmaint.sh  -i
This script will install libraries needed by genmon. This script assumes you have already downloaded the genmon project via 'git'. This script will optionally copy the config files to the configuration directory. This script will not test the serial port. This script requires internet access to download the needed libraries. Press any key to continue.  
Installing genmon package requirements....
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: Flask in /usr/lib/python3/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 1)) (1.1.2)
Requirement already satisfied: configparser in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 2)) (6.0.0)
Requirement already satisfied: pyserial in /usr/lib/python3/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 3)) (3.5b0)
Requirement already satisfied: crcmod in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 4)) (1.7)
Requirement already satisfied: pyowm==2.10.0 in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 5)) (2.10.0)
Requirement already satisfied: pytz in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 6)) (2023.3.post1)
Requirement already satisfied: pyopenssl in /usr/lib/python3/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 7)) (20.0.1)
Requirement already satisfied: twilio in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 8)) (8.11.1)
Requirement already satisfied: chump in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 9)) (1.6.0)
Requirement already satisfied: paho-mqtt in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 10)) (1.6.1)
Requirement already satisfied: smbus in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 11)) (1.1.post2)
Requirement already satisfied: pysnmp in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 12)) (4.4.12)
Requirement already satisfied: ldap3 in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 13)) (2.9.1)
Requirement already satisfied: pyasn1==0.4.8 in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 14)) (0.4.8)
Requirement already satisfied: pyotp in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 15)) (2.9.0)
Requirement already satisfied: psutil in /usr/lib/python3/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 17)) (5.8.0)
Requirement already satisfied: spidev in /usr/lib/python3/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 18)) (3.5)
Requirement already satisfied: bleson in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 19)) (0.1.8)
Requirement already satisfied: fluids in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 20)) (1.0.25)
Requirement already satisfied: voipms in /usr/local/lib/python3.9/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 21)) (0.2.5)
Requirement already satisfied: RPi.GPIO in /usr/lib/python3/dist-packages (from -r /home/cvin-admin/genmon/requirements.txt (line 22)) (0.7.0)
Requirement already satisfied: geojson<3,>=2.3.0 in /usr/local/lib/python3.9/dist-packages (from pyowm==2.10.0->-r /home/cvin-admin/genmon/requirements.txt (line 5)) (2.5.0)
Requirement already satisfied: requests<3,>=2.20.0 in /usr/lib/python3/dist-packages (from pyowm==2.10.0->-r /home/cvin-admin/genmon/requirements.txt (line 5)) (2.25.1)
Requirement already satisfied: scipy>=0.9.0 in /usr/local/lib/python3.9/dist-packages (from fluids->-r /home/cvin-admin/genmon/requirements.txt (line 20)) (1.12.0)
Requirement already satisfied: numpy>=1.5.0 in /usr/local/lib/python3.9/dist-packages (from fluids->-r /home/cvin-admin/genmon/requirements.txt (line 20)) (1.26.3)
Requirement already satisfied: pysmi in /usr/local/lib/python3.9/dist-packages (from pysnmp->-r /home/cvin-admin/genmon/requirements.txt (line 12)) (0.3.4)
Requirement already satisfied: pycryptodomex in /usr/local/lib/python3.9/dist-packages (from pysnmp->-r /home/cvin-admin/genmon/requirements.txt (line 12)) (3.20.0)
Requirement already satisfied: aiohttp-retry>=2.8.3 in /usr/local/lib/python3.9/dist-packages (from twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (2.8.3)
Requirement already satisfied: PyJWT<3.0.0,>=2.0.0 in /usr/local/lib/python3.9/dist-packages (from twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (2.8.0)
Requirement already satisfied: aiohttp>=3.8.4 in /usr/local/lib/python3.9/dist-packages (from twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (3.9.1)
Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (1.4.1)
Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (1.3.1)
Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (23.2.0)
Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (1.9.4)
Requirement already satisfied: async-timeout<5.0,>=4.0 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (4.0.3)
Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.9/dist-packages (from aiohttp>=3.8.4->twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (6.0.4)
Requirement already satisfied: idna>=2.0 in /usr/lib/python3/dist-packages (from yarl<2.0,>=1.0->aiohttp>=3.8.4->twilio->-r /home/cvin-admin/genmon/requirements.txt (line 8)) (2.10)
Requirement already satisfied: ply in /usr/local/lib/python3.9/dist-packages (from pysmi->pysnmp->-r /home/cvin-admin/genmon/requirements.txt (line 12)) (3.11)
Installing rpirtscts...
~/genmon ~/genmon
fatal: destination path 'rpirtscts' already exists and is not an empty directory.
make: 'rpirtscts' is up to date.
Finished installing rpirtscts.
~/genmon
Copy configuration files to /etc/genmon/? (y/n)?y
Copying *.conf files to /etc/genmon/
mkdir: cannot create directory ‘/etc/genmon/’: File exists
What type of connection from genmon to the controller? S=Onboard Serial, T=Network, Serial over TCP/IP Bridge, U=USB Serial (s/t/u)?s
Setting up serial onboard port...
Setup the raspberry pi onboard serial port? (y/n)?y
Setting up serial port...
~/genmon ~/genmon

Model: Raspberry Pi 4 Model B Rev 1.5
Backup file already exist: /boot/config.txt.bak1
Backup file already exist: /boot/cmdline.txt.bak1

Enable UART: Success
Change BT serial port: Success
Enable Turbo clocking: Success
Disable BT Disable Overlay: Success
Disable serial console: Success
Disable serial console service: Success
Created symlink /etc/systemd/system/dev-serial1.device.wants/hciuart.service → /lib/systemd/system/hciuart.service.
Enable BT service service: Success

Serial port settings are OK. A reboot is needed if changes were made.

Finished setting up the serial port.
~/genmon
Done.
Start genmon on boot? (y/n)?y
Updating crontab....
Crontab already contains genmon start script:
@reboot sleep 30 && /bin/bash /home/cvin-admin/genmon/startgenmon.sh start
cvin-admin@genmon:~/genmon $ cd OtherApps/
cvin-admin@genmon:~/genmon/OtherApps $ sudo python3 serialtest.py 

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5b0
Serial port opened
write data: sent test string
waiting to received data....
FAILED: Sent data does not match receive. Received 0 bytes
cvin-admin@genmon:~/genmon/OtherApps $ sudo python serialtest.py  

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5b0
Serial port opened
write data: sent test string
waiting to received data....
FAILED: Sent data does not match receive. Received 0 bytes
cvin-admin@genmon:~/genmon/OtherApps $ cd ..
cvin-admin@genmon:~/genmon $ bash ./startgenmon.sh -hardstop
Error: Unsupported flag -hardstop
cvin-admin@genmon:~/genmon $ bash ./startgenmon.sh 
Invalid command. Valid commands are start, stop, restart or hardstop.
cvin-admin@genmon:~/genmon $ bash ./startgenmon.sh hardstop 
Hard Stopping genmon python scripts
Running one time maintenance check
Stopping....
cvin-admin@genmon:~/genmon $ sudo python OtherApps/serialtest.py 

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5b0
Serial port opened
write data: sent test string
waiting to received data....
FAILED: Sent data does not match receive. Received 0 bytes
```

**Comment by jgyates** [comment]
Is the above serailtest with or without the HAT? If not hat then send a picture of how your are connecting pin 8 to pin 10.

**Reply by jgyates** [reply]
sending a pick of how your are looping back may help if there is any confusion on the pin numbering.

**Reply by EvanET13** [reply]
I am looping without the HAT. 
![IMG_8903](https://github.com/jgyates/genmon/assets/60356533/fef38927-4487-471f-8735-ee390a749031)
![IMG_8902](https://github.com/jgyates/genmon/assets/60356533/e80e8f4b-8a0a-491d-812a-02f8af61e1a8)
![IMG_8904](https://github.com/jgyates/genmon/assets/60356533/6aaa017b-b403-4e02-ae3b-f385a7e3d256)

Using pins 8 and 10, connecting them together via a jumper wire

**Reply by jgyates** [reply]
one thing that is a bit odd is this line in the output of serialtest.py:

     Using python serial library V3.5b0

This is a beta version of the serial driver. I have seen this happen once before. Apparently this happens if a mirror server is improperly configured and that mirror server is selected in your install process. The version should be V3.5. Try this:

     sudo pip3 install pyserial --upgrade --force-reinstall

Then retry the loopback test.

**Reply by jgyates** [reply]
if that does not work, then try this with serailtest.py

      python3 serialtest.py /dev/serial1

**Reply by EvanET13** [reply]
After running ` sudo pip3 install pyserial --upgrade --force-reinstall` : 
![image](https://github.com/jgyates/genmon/assets/60356533/f68fb4e9-8896-448c-b3d1-5e79d65373b6)

**Reply by EvanET13** [reply]
When running as `  python3 serialtest.py /dev/serial1` : 
![image](https://github.com/jgyates/genmon/assets/60356533/c5d5cb85-23d2-4e77-a1b0-6f8e2aa7ea32)

**Reply by jgyates** [reply]
what are the last lines 4 lines in /boot/config.txt

also show the output of /boot/cmdline.txt

**Reply by EvanET13** [reply]
Last 4 lines of /boot/config.txt:
```
[all]
dtoverlay=gpio-fan,gpiopin=14,temp=80000
dtoverlay=miniuart-bt                 // <---- is this the issue?
force_turbo=1
```

Output of /boot/cmdline.txt:
`console=tty1 root=PARTUUID=cfca88d0-02 rootfstype=ext4 fsck.repair=yes rootwait quiet splash plymouth.ignore-serial-consoles`

**Reply by jgyates** [reply]
I am not seeing the line "enable_uart"=1. Here are the last four lines of my pi4 system

     [all]
     enable_uart=1
     dtoverlay=miniuart-bt
     force_turbo=1

The miniuart-bt tells the Bluetooth software to use the mini UART and not the full featured UART in the pi SoC. That is expected if you want to use BT and the serial port for genmon at the same time

Did you run raspi-config after running serailconfig.py? I wonder where the gpio-fan stuff came from?

**Reply by EvanET13** [reply]
It's weird because I added the `enable_uart=1` manually and that didnt work. So I tried the `serialconfig.py -e` then when the check; with passing results.

Should I manually add `enable_uart=1` into /boot/config.txt? Then run the serialtest.py?

>>Did you run raspi-config after running serailconfig.py? I wonder where the gpio-fan stuff came from?
raspi-config was done before serialconfig.py. gpio-fan stuff comes from an additional HAT that I have that allows for PoE (HAT has a fan on it)

**Reply by EvanET13** [reply]
I do have `enable_uart=1` just was a few lines up
![image](https://github.com/jgyates/genmon/assets/60356533/fe3b9d05-dc0d-4250-981b-7d891c2e0a7b)

**Reply by jgyates** [reply]
@EvanET13, there are only a few more things to try:

1) try to consolidate the [all] sections into one in your /boot/config.txt file. Reboot and see if that changes the behavior
2) Reinstall the OS and try again
3) try a different pi

This is the most baseline test of the serial port. The only thing that is a variable is the config.txt and cmdline.txt file. Presently my two best guesses are: There are either conflicting settings in your config file, or your pi is bad.

If you reinstall, then run the instructions on [this page](https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software) first. Then test the serial port 2nd. This will minimize the changes to config.txt to see if some other settings is conflicting with the ones that serialconfig.py is making.

Let us know what happens.

**Reply by EvanET13** [reply, marked as answer]
@jgyates I got it working. I reformatted my pi. This time I got to choose the OS and chose the 64-bit version; it came preloaded with 32-bit version.
In addition, I did the following differently:
1. I did not connect my PoE HAT initially (still haven't)
2. Established `enable_uart=1` in config.txt before  `git clone https://github.com/jgyates/genmon.git` Then did the `serialtest.py`

Ran the ` bash ./genmonmaint.sh -i`  -> Ran the test again, working.

**Reply by jgyates** [reply]
Great, I am glad it is working for you. Thanks for lettings everyone know the solution.

**Comment by jgyates** [comment]
also, what version of the OS are you running? Type this to find out:

     cat /etc/os-release

**Reply by EvanET13** [reply]
![image](https://github.com/jgyates/genmon/assets/60356533/4ab52f6e-5570-40ca-b917-4b97552c6e0f)

**Comment by liltux** [comment]
@EvanET13 Can you verify your Evolution controller firmware?  On the main system screen (System Date/Time ... Submenus) of the Evolution controller press the button sequence Escape, Up, Up, Down, Down, Escape, Escape.  You will then be able to navigate to the Dealer menus (Submenu > Dealer).  In Dealer menu, select Display, and using the up/down arrow navigate to the firmware screen.  One of the older/original firmware of this controller used a different baud rate and a firmware update will correct this issue.  Just throwing that out there.

**Reply by EvanET13** [reply]
@liltux Thank you for that information. I will check this out the next time I go out to my site with this generator. For now I'm making sure I have the pi configured properly first.

**Comment by skipfire** [comment]
In your screenshots I see you have a non-standard username, was that renamed in the imager, or did you create that user after?  I think there is a permission around the serial port and a new user may not be in that group, even if it is an admin user.
